core/common/src/main/java/alluxio/security/authentication/ChannelAuthenticator.java

```diff
@@ -0,0 +1,148 @@
+package alluxio.security.authentication;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.exception.status.UnauthenticatedException;
+import alluxio.grpc.SaslAuthenticationServiceGrpc;
+import alluxio.grpc.SaslMessage;
+import alluxio.util.SecurityUtils;
+import alluxio.grpc.GrpcChannelBuilder;
+import io.grpc.Channel;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors;
+import io.grpc.ManagedChannel;
+import io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.StreamObserver;
+
+import javax.security.auth.Subject;
+import javax.security.sasl.AuthenticationException;
+import javax.security.sasl.SaslClient;
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Used to authenticate with the target host. Used internally by {@link GrpcChannelBuilder}.
+ */
+public class ChannelAuthenticator {
```

Comment:
I'm not clear on why this class needs state, could this be a util class ?

Reply:
It needs a state on what kind of user information to use. Subject vs plain username/password.
It could easily be an util class, but with inflated method signatures.